core/drivers/crypto/marvell/cipher.c

```diff
@@ -0,0 +1,255 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2025, Marvell
+ */
+
+#include <assert.h>
+#include <crypto/crypto.h>
+#include <crypto/crypto_impl.h>
+#include <crypto/internal_aes-gcm.h>
+#include <drvcrypt.h>
+#include <drvcrypt_cipher.h>
+#include <initcall.h>
+#include <stdlib.h>
+#include <string.h>
+#include <string_ext.h>
+#include <tee_api_types.h>
+#include <tee/cache.h>
+#include <utee_defines.h>
+#include <util.h>
+
+#include "mrvl_ehsm_cryp.h"
+
+struct mrvl_cipher_ctx {
+	struct crypto_cipher_ctx a_ctx;
+	struct mrvl_cryp_context cryp_ctx;
+	enum mrvl_cryp_algo_mode algo;
+};
+
+static struct mrvl_cipher_ctx *to_mrvl_cipher_ctx(struct crypto_cipher_ctx *ctx)
+{
+	assert(ctx);
+
+	return container_of(ctx, struct mrvl_cipher_ctx, a_ctx);
+}
+
+static TEE_Result mrvl_cipher_initialize(struct drvcrypt_cipher_init *dinit)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	struct mrvl_cipher_ctx *c = to_mrvl_cipher_ctx(dinit->ctx);
+	struct mrvl_cryp_context *ctx = &c->cryp_ctx;
+
+	mrvl_ehsm_cryp_lock();
+
+	// if already initialized free the previous content and context
+	if (ctx->initialized) {
+		free(ctx->iv_data);
+		free(ctx->key);
+		if (ctx->key2)
+			free(ctx->key2);
+
+		if (ctx->context_id != 0) {
+			ret = mrvl_ehsm_aes_context_release(ctx->context_id);
+			if (ret != TEE_SUCCESS)
+				goto out;
+		}
+	}
+
+	ctx->iv_len = dinit->iv.length;
+	ctx->iv_data = mem_alloc(ctx->iv_len);
+	if (!ctx->iv_data) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+	memcpy(ctx->iv_data, dinit->iv.data, ctx->iv_len);
+
+	ctx->op_enc_dec = dinit->encrypt;
+
+	ctx->key_len = dinit->key1.length;
+	ctx->key = mem_alloc(ctx->key_len);
+	if (!ctx->key) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out_key_err;
+	}
+	memcpy(ctx->key, dinit->key1.data, ctx->key_len);
+
+	if (dinit->key2.length) {
+		ctx->key2_len = dinit->key2.length;
+		ctx->key2 = mem_alloc(ctx->key2_len);
+		if (!ctx->key2) {
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto out_key2_err;
+		}
+		memcpy(ctx->key2, dinit->key2.data, ctx->key2_len);
+	}
+
+	ret = mrvl_ehsm_aes_init(c->algo, ctx->op_enc_dec, (void *)ctx->key,
+				 ctx->key_len, (void *)ctx->key2, ctx->key2_len,
+				 ctx->iv_data, 0);
+	if (ret != TEE_SUCCESS)
+		goto out_init_err;
+
+	ret = mrvl_ehsm_aes_context_store(&ctx->context_id);
+	if (ret != TEE_SUCCESS)
+		goto out_init_err;
+
+	ctx->is_new = true;
+	ctx->initialized = true;
+
+	goto out;
+
+out_init_err:
+	if (ctx->key2)
+		free(ctx->key2);
+out_key2_err:
+	free(ctx->key);
+out_key_err:
+	free(ctx->iv_data);
+out:
+	mrvl_ehsm_cryp_unlock();
+	return ret;
```

Comment:
To simplify such error and out path (here and elsewhere in this P-R), I would suggest:
```c
out:
	if (ret != TEE_SUCCESS) {
		free(ctx->key2);
		free(ctx->key);
		free(ctx->iv_data);
	}

	mrvl_ehsm_cryp_unlock();

	return ret;